### PR TITLE
RNA I/O module

### PR DESCRIPTION
**Pull request by joergi-w**

### CODE REVIEW

File: tests/rna_io/test_rna_io.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include <seqan/file.h>
+
+#include "test_rna_io.h"
+
+
+SEQAN_BEGIN_TESTSUITE(test_rna_io)
+{
+    // Call tests.
+    SEQAN_CALL_TEST(test_rna_io_read_connect);
+	SEQAN_CALL_TEST(test_rna_io_read_dot_bracket);
```
Comment (smehringer) [inline tests/rna_io/test_rna_io.cpp]: is the indentation done on purpose?

File: include/seqan/rna_io/bpseq_read_write.h [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+// --------------------------------------------------------------------------
+
+template <typename T>
+struct FileExtensions<Bpseq, T>
+{
+    static char const * VALUE[1];    // default is one extension
+};
+
+template <typename T>
+char const * FileExtensions<Bpseq, T>::VALUE[1] =
```
Comment (smehringer) [inline include/seqan/rna_io/bpseq_read_write.h]: You probably have to model that as a constexpr instead?

File: include/seqan/rna_io/vienna_read_write.h [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    if (pos == std::string::npos)
+    {
+        record.name = buffer;
+    }
+    else
+    {
+        record.name = buffer.substr(0, pos);
+        std::string posStr{buffer.substr(pos + 1)};
+        pos = posStr.find('-');
+        if (pos == std::string::npos || !lexicalCast(record.offset, posStr.substr(0, pos))) {
```
Comment (joergi-w) [inline include/seqan/rna_io/vienna_read_write.h]: Thank you for the hint!

File: include/seqan/rna_io/bpseq_read_write.h (hunk 2 above)
Comment (joergi-w) [inline include/seqan/rna_io/bpseq_read_write.h]: I do not think that constexpr is possible here, as the variable must live in the FileExtensions struct. I guess this is necessary for `guessFormatFromFilename`?

File: include/seqan/rna_io/rna_record.h [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+    /*!
+     * @fn RnaRecord::RnaRecord
+     * @brief The constructor.
+     * @signature RnaRecord::RnaRecord()
+     */
+    RnaRecord() : recordID(UNDEF), seqLen(0u), offset(1u), name(""), quality(""),
+                  comment(""), sequence("")
+    {}
+
```
Comment (rrahn) [inline include/seqan/rna_io/rna_record.h]: Use default member-initializer for offset and remove the def-constructor.
Default aggregate initialization should do the rest.

File: include/seqan/rna_io/rna_struct_file.h [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+ * @see FileFormats#Bpseq
+ * @see FileFormats#Ebpseq
+ *
+ * This tag is used as TSpec template parameter in @link FormattedFile @endlink.
+ */
+struct RnaStruct_;
+typedef Tag<RnaStruct_> RnaStruct;
+
+typedef
+TagList<Connect,
```
Comment (rrahn) [inline include/seqan/rna_io/rna_struct_file.h]: please indent the first TagList and put all others beneath on the same position.

File: include/seqan/rna_io/rna_struct_file.h [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    std::vector<RnaRecord> records;
+
+    /*!
+     * @var RnaHeader RnaStructContents::header
+     * @brief The (pseudo) header of an RNA structure file.
+     * @see RnaHeader
+     */
+    RnaHeader header;
+};
+typedef struct RnaStructContents_ RnaStructContents;
```
Comment (rrahn) [inline include/seqan/rna_io/rna_struct_file.h]: why redefining RnaStructContents_ as RnaStructContents?

File: include/seqan/rna_io/rna_struct_file.h [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+{
+    typedef typename TTagList::Type TFormat;
+
+    if (isEqual(format, TFormat()))
+        writeRecord(iter, record, context, TFormat());
+    else
+        writeRecord(iter, record, context, static_cast<typename TagSelector<TTagList>::Base const &>(format));
+}
+
+
```
Comment (rrahn) [inline include/seqan/rna_io/rna_struct_file.h]: double new-line

File: include/seqan/rna_io/rna_struct_file.h [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+ * @throw IOError On low-level I/O errors.
+ * @throw ParseError On high-level file format errors.
+ */
+template <typename TSpec>
+inline void
+readHeader(RnaHeader & header, FormattedFile<RnaStruct, Input, TSpec> & file)
+{
+    if (isEqual(file.format, Ebpseq()))
+        readHeader(header, context(file), file.iter, Ebpseq());
+    // other files contain no header
```
Comment (rrahn) [inline include/seqan/rna_io/rna_struct_file.h]: put comment behind the if statement

File: include/seqan/rna_io/rna_struct_file.h [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+ * @throw IOError On low-level I/O errors.
+ * @throw ParseError On high-level file format errors.
+ */
+template <typename TSpec>
+inline void
+writeHeader(FormattedFile<RnaStruct, Output, TSpec> & file, RnaHeader const & header)
+{
+    if (isEqual(file.format, Ebpseq()))
+        writeHeader(file.iter, header, context(file), Ebpseq());
+    // other files contain no header
```
Comment (rrahn) [inline include/seqan/rna_io/rna_struct_file.h]: same here

File: include/seqan/rna_io/stockholm_read_write.h [hunk 10]
```diff
@@ -0,0 +1,333 @@
+// ==========================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ==========================================================================
+// Copyright (c) 2006-2015, Knut Reinert, FU Berlin
```
Comment (rrahn) [inline include/seqan/rna_io/stockholm_read_write.h]: Please update the copyright year and possibly in all other files.

File: include/seqan/rna_io/stockholm_read_write.h [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+        }
+        skipUntil(iter, IsNewline());
+        skipOne(iter);
+        clear(buffer);
+    }
+    if (empty(bracketStr))
+        SEQAN_THROW(ParseError("Expected a secondary structure line."));
+    if (empty(sequence))
+        SEQAN_THROW(ParseError("Expected a sequence line."));
+    SEQAN_ASSERT_EQ(length(sequence), length(gapPos));
```
Comment (rrahn) [inline include/seqan/rna_io/stockholm_read_write.h]: put new line before

File: include/seqan/rna_io/stockholm_read_write.h [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+
+    // store the alignment with gaps in record
+    resize(rows(record.align), length(sequence));
+    for (unsigned seq = 0; seq < length(sequence); ++seq)
+    {
+        assignSource(row(record.align, seq), sequence[seq]);
+        for (unsigned pos = length(gapPos[seq]); pos > 0u; --pos)
+            insertGap(row(record.align, seq), gapPos[seq][pos - 1]);
```
Comment (rrahn) [inline include/seqan/rna_io/stockholm_read_write.h]: brackets

File: include/seqan/rna_io/vienna_read_write.h [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+            }
+            else
+            {
+                SEQAN_THROW(ParseError("Invalid bracket notation: unpaired closing bracket"));
+            }
+        }
+    }
+    if(!stack.empty())
+        SEQAN_THROW(ParseError("Invalid bracket notation: unpaired opening bracket"));
+    append(record.fixedGraphs, graph);
```
Comment (rrahn) [inline include/seqan/rna_io/vienna_read_write.h]: add new-line before

File: include/seqan/rna_io/vienna_read_write.h [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    clear(buffer);
+
+    // read energy if present
+    skipUntil(iter, OrFunctor<IsNewline, EqualsChar<'('> >());
+    if (!atEnd(iter) && value(iter) == '(')
+    {
+        skipOne(iter);
+        readUntil(buffer, iter, EqualsChar<')'>());
+        if (!lexicalCast(record.fixedGraphs[0].energy, buffer))
+            SEQAN_THROW(BadLexicalCast(record.fixedGraphs[0].energy, buffer));
```
Comment (rrahn) [inline include/seqan/rna_io/vienna_read_write.h]: brackets

File: include/seqan/rna_io/vienna_read_write.h [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+        }
+        else                                            // close bracket
+        {
+            bracketStr[idx] = ')';
+            if (stack.empty())
+                SEQAN_FAIL("Cannot reach here.");
+            if (stack.top() == idx)
+                stack.pop();
+            else
+                SEQAN_THROW(ParseError("ERROR: Vienna format does not allow pseudoknots."));
```
Comment (rrahn) [inline include/seqan/rna_io/vienna_read_write.h]: brackets to the block above

File: tests/rna_io/test_rna_io.cpp [hunk 16]
```diff
@@ -0,0 +1,57 @@
+// ==========================================================================
+//                                   rna_io
```
Comment (rrahn) [inline tests/rna_io/test_rna_io.cpp]: correct header format. Also check all other test header files

File: tests/rna_io/test_rna_io.h [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+    // set sequence identifiers
+    seqan::appendValue(record.seqID, seqan::CharString{"seq0"});
+    seqan::appendValue(record.seqID, seqan::CharString{"seq1"});
+
+    seqan::RnaStructureGraph graph;
+    for (unsigned idx = 0; idx < record.seqLen; ++idx)
+        addVertex(graph.inter);
+    for (unsigned idx = 0; idx < 4; ++idx)
+        addEdge(graph.inter, idx, 7u - idx, 1.);
+    append(record.fixedGraphs, graph);
```
Comment (rrahn) [inline tests/rna_io/test_rna_io.h]: new line before

File: tests/rna_io/test_rna_io.h [hunk 18]
```diff
@@ -0,0 +1,444 @@
...
+    record.name = "S.cerevisiae_tRNA-PHE";
+    record.sequence = "GCGGAUUU";
+    record.seqLen = static_cast<unsigned>(length(record.sequence));
+    seqan::RnaStructureGraph graph;
+
+    for (unsigned idx = 0; idx < record.seqLen; ++idx)
+        addVertex(graph.inter);
+    for (unsigned idx = 0; idx < 4; ++idx)
+        addEdge(graph.inter, idx, 7u - idx, 1.);
+    append(record.fixedGraphs, graph);
```
Comment (rrahn) [inline tests/rna_io/test_rna_io.h]: newline before

File: include/seqan/rna_io/bpseq_read_write.h [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+        skipLine(iter);
+        skipUntil(iter, NotFunctor<IsWhitespace>());
+    }
+    append(record.fixedGraphs, graph);
+    record.seqLen = currPos;  //set amount of records
+}
+
+template <typename TForwardIter>
+inline void
+readRecord(RnaRecord & record, SEQAN_UNUSED RnaIOContext &, TForwardIter & iter, Bpseq const & /*tag*/)
```
Comment (joergi-w) [inline include/seqan/rna_io/bpseq_read_write.h]: I thought the macro was meant to indicate function arguments which are unused on purpose... Now I delete the macro everywhere I used it like this.

File: include/seqan/rna_io/stockholm_read_write.h [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+char const * FileExtensions<Stockholm, T>::VALUE[1] =
+{
+    ".sth"      // default output extension
+};
+
+// ==========================================================================
+// Functions
+// ==========================================================================
+
+
```
Comment (rrahn) [inline include/seqan/rna_io/stockholm_read_write.h]: double new-line

**Comment by rrahn** [comment]
OK looks good to me!

**Pull request merged**